extras/module-node/duk_module_node.c

```diff
@@ -0,0 +1,263 @@
+
+ have_module:
+	/* [ ... module ] */
+
+	(void) duk_get_prop_string(ctx, -1, "exports");
+	return 1;
+}
+
+static void duk__push_require_function(duk_context *ctx, const char *id) {
+	duk_push_c_function(ctx, duk__handle_require, 1);
+	duk_push_string(ctx, "name");
+	duk_push_string(ctx, "require");
+	duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE);
+	duk_push_string(ctx, id);
+	duk_put_prop_string(ctx, -2, "\xffmoduleID");
+
+	/* require.cache */
+	duk_push_global_stash(ctx);
+	(void) duk_get_prop_string(ctx, -1, "\xffrequireCache");
+	duk_put_prop_string(ctx, -3, "cache");
+	duk_pop(ctx);
+}
+
+static void duk__push_module_object(duk_context *ctx, const char *id) {
+	duk_push_object(ctx);
+
+	/* Node.js uses the canonicalized filename of a module for both module.id
+	 * and module.filename.  We have no concept of a file system here, so just
+	 * use the module ID for both values.
+	 */
+	duk_push_string(ctx, id);
+	duk_dup(ctx, -1);
+	duk_put_prop_string(ctx, -3, "filename");
+	duk_put_prop_string(ctx, -2, "id");
+
+	/* module.exports = {} */
+	duk_push_object(ctx);
+	duk_put_prop_string(ctx, -2, "exports");
+
+	/* module.loaded = false */
+	duk_push_false(ctx);
+	duk_put_prop_string(ctx, -2, "loaded");
+
+	/* module.require */
+	duk__push_require_function(ctx, id);
+	duk_put_prop_string(ctx, -2, "require");
+}
+
+static duk_int_t duk__eval_module_source(duk_context *ctx) {
+	/*
+	 *  Stack: [ ... module source ] => [ ... module ] (retval == 0)
+	 *         [ ... module source ] => [ ... err ]    (retval != 0)
+	 */
+
+	/* Wrap the module code in a function.  This is the easiest way to
+	 * implement CommonJS closure semantics and matches the behavior of e.g.
+	 * Node.js.
+	 */
+	duk_push_string(ctx, "(function main(exports, require, module, __filename) { ");
+	duk_dup(ctx, -2);  /* source */
+	duk_push_string(ctx, " })");
+	duk_concat(ctx, 3);
+
+	/* [ ... module source func_src ] */
+
+	duk_remove(ctx, -2);
+
+	/* [ ... module func_str ] */
+
+	(void) duk_get_prop_string(ctx, -2, "filename");
+	if (duk_pcompile(ctx, DUK_COMPILE_EVAL) != DUK_EXEC_SUCCESS) {
+		duk_remove(ctx, -2);  /* [ ... err ] */
+		goto fail;
+	}
+	duk_call(ctx, 0);
+
+	/* [ ... module func ] */
+
+	/* call the function wrapper */
+	(void) duk_get_prop_string(ctx, -2, "exports");   /* exports */
+	(void) duk_get_prop_string(ctx, -3, "require");   /* require */
+	duk_dup(ctx, -4);                                 /* module */
+	(void) duk_get_prop_string(ctx, -5, "filename");  /* __filename */
+	if (duk_pcall(ctx, 4) != DUK_EXEC_SUCCESS) {
+		duk_remove(ctx, -2);  /* [ ... err ] */
+		goto fail;
+	}
+	duk_pop(ctx);
+
+	/* [ ... module ] */
+
+	duk_push_true(ctx);
+	duk_put_prop_string(ctx, -2, "loaded");
+
+	return DUK_EXEC_SUCCESS;
+
+ fail:
+	/* [ ... err ] */
+
+	return DUK_EXEC_ERROR;
+}
+
+void duk_module_node_init(duk_context *ctx) {
+	/*
+	 *  Stack: [ ... options ] => [ ... ]
+	 */
+
+	duk_idx_t options_idx;
+
+	duk_require_object_coercible(ctx, -1);
```

Comment:
Hmm, well, `\xffrequireCache` gets written before that so maybe this is useful after all.